### DISCUSSION TITLE
Argument table

### DISCUSSION THREAD

**Discussion by waldo1001** [microsoft/alguidelines#97] [BC Patterns]
Very much like the previous one, but include extensibility, tabletype temporary, ..

**Comment by StefanMaron** [comment]
Why use a table in the first place when we have dictionaries and lists now?

Coming from python, I know that there you have this feature even as a fixed part of the language:
https://book.pythontips.com/en/latest/args_and_kwargs.html

I can not recall where I have read it, but someone testet it, and temp tables are significantly slower then lists or dictionaries.

**Comment by waldo1001** [comment]
Fair question - and this is what we came up with: "tabletype temp" use the same mechanics as lists/dictionaries, and is a bit more convenient to work with: table extensions, different datatypes, ... .

But if you say now that it's slower - may be it deserves a test ;-).

**Reply by StefanMaron** [reply]
ha, found it. Its mentioned in the bcperformance article 
https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/performance/performance-developer#builtindatastructure

![image](https://user-images.githubusercontent.com/24838311/152139688-48f324c4-1ae0-4dda-9d3d-7412959b88f4.png)

**Comment by waldo1001** [comment]
Yes, but it gets a bit more complicated when you need to mix different datatypes.. which in an argument-table-situation isn't that unthinkable..

**Reply by StefanMaron** [reply]
I see, I did not think about that. Converting everything to text is not the best solution :D
List of variant does not even work

**Reply by StefanMaron** [reply]
But then again, the dictionary approach is already used for page background tasks

**Reply by StefanMaron** [reply]
Maybe JsonObjects and JsonArrays?  
![image](https://user-images.githubusercontent.com/24838311/152141155-e52c8c61-a4b4-4783-bb22-c95c4685575b.png)

**Reply by waldo1001** [reply]
That could be considered for sure imho ..

**Reply by klauskaan** [reply]
I love your approach Stefan! I would like to see it in action.

On the other hand, i would still like to see the argument table pattern described as it is sure to be widely implemented .

**Comment by JeremyVyska** [comment]
Shared this one out to Twitter to try to get some eyes on it.

I actually _really_ like the JsonObject strategy as a way of passing a variety of complex data types in a Dictionary way.

I'm very curious to see what folks think about the Argument Table vs JsonObject strategy, which I think are the two most comparable.  I think the real differentiator is:   With an Argument Table, you define the Keys of the 'dictionary' ahead of time, so they're strongly typed, plus you get the syntax sugar of a table - you can define validation logic, easily default values, and check table relationships.     I don't say that's a pro or a con, just a differentiator.  If you need none of those 'sugar' pieces, then JsonObject is pretty _sweet_ too.

**Reply by StefanMaron** [reply]
Not sure if I would want to have any of this validation inside the table tho. I would much rather see it within the function or as a separate sub function which does validation of the parameters. 

For me that would result in cleaner and more understandable code 🤔

**Reply by ajkauffmann** [reply]
A table is a complex type, even when TableType = Temporary. It comes close to a class, but still carries a lot overhead from the table object.
Table extensibility could be an argument (pun intended), but with JSON objects you can easily achieve the same with proper events.
On the contrary, a JSON object always requires extra code (or helper functions) to read data. Not a big deal though.
But JSON has other advantages, like optional parameters. If it is not there, then it is not defined. Try that with a table object where all fields always exist.
So after all, I believe I'm in favor of JSON as parameter objects.

**Reply by StefanMaron** [reply]
I like the idea to think about tables like classes in general already, and here again its a perfect fit. If the table holds different information which belongs together (aka a whole object) and those should be passed into a function. Then a table makes perfect sense. 

For example I need multiple values from a certain customer, then I pass in the whole customer as a record. 

If the parameter table is a construct like this, go for the table not for JSON.
But if the fields of the table would just be a collection of independent values, JSON would be better.

**Reply by pkoutsogilas** [reply]
Or create a generic "parameter handler" codeunit for working the JSON object or would that also increase the footprint / handling?

**Reply by waldo1001** [reply]
Personally, I don't think we should look into the performance-part, as ANY performance impact will be negligible in my honest opinion.
The only thing we should look at is "readability", and if the abstract JSON against the self-documenting-temptable are the two choices - I'd go for a table any time - intellisense will show me what all arguments are 🤷‍♂️.

**Reply by pkoutsogilas** [reply]
You indeed have to weigh the pros and cons of the solutions...

**Reply by vhn** [reply]
This seems like the good old dynamic types vs static types debate, but using BC concepts.
With temptable (static types) comes intellisense, compiler checks during later refactorings, compiler checks for dependencies, self-documenting code etc.
I personally much prefer this for the same reason I prefer typescript over javascript or C# over python -> because I love when my developer tools help me catch bugs before runtime errors happen. 

Whenever this typing debate comes up in any language ecosystem, it seems to split the waters a lot - some developers seem to hate being forced to think about & declare their data contracts up front, preferring to "write it out as they go a long". I also often hear arguments such as "Dynamic types increases developer velocity for the 1st version/PoC". In my own experience, I'd say that the PoC tends to become the real thing before cleaning up, and/or that it also increases the velocity of bugs.. But it's clearly subjective so my guess is that both versions of the pattern being described here will end up in production.😊

Apart from that typing debate, the fact that table arguments can be passed to the error handling OnRun codeunit trigger in AL is one big plus it has over any over types such as JsonObject (sadly.. I wish MS would expose the OnRun behavior in an attribute i.e. [TryFunctionWithTransactions] or whatever name, so this wasn't a thing to consider).

**Reply by pkoutsogilas** [reply]
I personally also prefer the Table (SubType = Temporary) method.

**Reply by StefanMaron** [reply]
I hear all of the arguments you have for the table version. 

But I thought the whole point of this would be to have a set of parameters which would be optional, without the need to define a whole lot of function overloads. In this case the function will need to validate this set of extra parameters in anyway to handle the optional part. 

If you do not want to have those set of parameters (those within the table) optional, then why create a table in the first place and not just use normal parameters.

Then the only advantage would be extensibility. But when the table is extended, the extra functionality will still need to be added via event. Via Json this can be done without the need for a table extension.

According runtime vs compile errors: Yes, if you go for json and have misspelled key names, you wont get any error during development. But with tests in place that would be absolutely no concern for me. And test would be necessary anyways to make sure this whole abstract parameter stuff works

**Reply by waldo1001** [reply]
I wasn't actually considering optional parameters .. 🤔.

And as extra - you can do Codeunit.Run with a table (where the table holds the arguments for the codeunit) - try that with a JSON ;-).
(--> less of an issue now we have interfaces though .. )

**Reply by StefanMaron** [reply]
Well, codeunit.run() would be a scenario where argument table has the advantage. And I am not saying that one or the other is the all time go to solution for this. 

The scenario I had in my head all the time actually is all about optional parameters. In what other scenario would I use a parameter table?

**Reply by pibcht** [reply]
> I like the idea to think about tables like classes in general already, and here again its a perfect fit. If the table holds different information which belongs together (aka a whole object) and those should be passed into a function. Then a table makes perfect sense.
> 
> For example I need multiple values from a certain customer, then I pass in the whole customer as a record.
> 
> If the parameter table is a construct like this, go for the table not for JSON. But if the fields of the table would just be a collection of independent values, JSON would be better.

I think Stefan touch the truth 🤔
I'd like to add two points:

1. Vanilla parameters are also valid answer
2. Don't forget currying

```al
procedure DoSomethingWithSmolStringList(s1: Text)
var
    ss: List of [Text];
begin
    ss.Add(s1);
    DoSomethingWithSmolStringList(ss);
end;

procedure DoSomethingWithSmolStringList(s1: Text; s2: Text)
var
    ss: List of [Text];
begin
    ss.Add(s1);
    ss.Add(s2);
    DoSomethingWithSmolStringList(ss);
end;

procedure DoSomethingWithSmolStringList(s1: Text; s2: Text; s3: Text)
var
    ss: List of [Text];
begin
    ss.Add(s1);
    ss.Add(s2);
    ss.Add(s3);
    DoSomethingWithSmolStringList(ss);
end;

procedure DoSomethingWithSmolStringList(ss: List of [Text])
begin
    <Do some magic>
end;
```

**Comment by mortenseifert** [comment]
I really like both. The table is strong typed and is clear to the caller what to do. Where as the Json is much more flexible and even allow for complex types, e.g. customer no. And list og item / qty. to create a sales order for.

I am thinking that the real ask is for Microsoft to support a struct (or even a class) type in AL. Or is this too much to ask for on a Monday 🤠

**Comment by gunnargestsson** [comment]
You can’t do Codeunit.Run() on List or Dictionary.  An Argument table has my vote.

**Comment by iti-witold-szal** [comment]
During Design Patterns training I was conducting, one of the young employees asked: "Why not use codeunit instead of a table? We could ad global var's and set/get methods like in other languages". I did not test this idea, but... What do you think?

**Reply by waldo1001** [reply]
I like it, but in my opinion, you kind of like have the same end result as a table when you're purely speaking about "arguments".

I do see codeunit-as-a-class possible in many other scenarios though.  I don't know how to call it, but @JesperSchulz and @SimonOfHH (https://github.com/SimonOfHH) used it in the AzureBlobStorage library in the system app.  It would make a perfect pattern ..  @JesperSchulz  @SimonOfHH ? ;-)

just an example of what I mean:
https://github.com/microsoft/ALAppExtensions/blob/main/Modules/System/Azure%20Blob%20Services%20API/src/ABSOperationPayload.Codeunit.al

**Reply by StefanMaron** [reply]
I could also imagine some snippets to generate those class like codeunits.
And while we are at it... How about some interfaces in the system app to have a common set of functions every class should have?

**Reply by SimonOfHH** [reply]
I also don't know how to call it 😅 For the Azure Blob Storage module it was kind of necessary to have something flexible at hand, because there are so many possible but not mandatory arguments. But I remember quite some hesitation and discussions in the beginning with @mazhelez when I started the first PR (because it was an uncommon way of doing things in BC). 

I think what speaks pro codeunit-approach is that there wouldn't be any kind of confusion in regards if it is schema-relevant or not. Also - as @StefanMaron mentions - you could work with interfaces to have a common approach present.

**Reply by mazhelez** [reply]
Is System Application we have example of both approaches — using a codeunit ([ABS Optional Parameters](https://github.com/microsoft/ALAppExtensions/blob/main/Modules/System/Azure%20Blob%20Services%20API/src/Helper/ABSOptionalParameters.Codeunit.al)) and using a purely temporary table ([Regex Options](https://github.com/microsoft/ALAppExtensions/blob/main/Modules/System/Regex/src/RegexOptions.Table.al)).

Generally, I'm in favor for the temporary table approach, as this is exactly what temporary tables are for: transferring data around, DTOs.
However, they have a huge flaw (the reason we went with a codeunit approach in ABS module). Namely, text fields are limited. 

Both approaches are fine, as long as the object only transfers data around and if there's any logic, it's only regarding the data it holds.

**Reply by waldo1001** [reply]
"Both approaches are fine, as long as the object only transfers data around and if there's any logic, it's only regarding the data it holds" - very important indeed! 👍
